contracts/bridge/BladeManager.sol

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+import "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../lib/GenesisLib.sol";
+import "../interfaces/bridge/IBladeManager.sol";
+import "../interfaces/bridge/IRootERC20Predicate.sol";
+
+contract BladeManager is IBladeManager, Ownable2StepUpgradeable {
+    using SafeERC20 for IERC20;
+    using GenesisLib for GenesisSet;
+
+    IRootERC20Predicate private _rootERC20Predicate;
+    GenesisSet private _genesis;
+
+    function initialize(address newRootERC20Predicate, GenesisAccount[] calldata genesisValidators) public initializer {
+        require(newRootERC20Predicate != address(0), "INVALID_INPUT");
+
+        _rootERC20Predicate = IRootERC20Predicate(newRootERC20Predicate);
+
+        uint256 length = genesisValidators.length;
+        for (uint256 i = 0; i < length; ++i) {
+            _genesis.insert(
+                genesisValidators[i].addr,
+                genesisValidators[i].nonStakedTokens,
+                genesisValidators[i].stakedTokens,
+                true
+            );
```

Comment:
Why not providing here non-validator genesis accounts as well and populate premine info?

```suggestion
                genesisValidators[i].isValidator
```

Reply:
Fix: https://github.com/Ethernal-Tech/blade-contracts/pull/9/commits/f2942acb1332ddeb385a3107648dafeba77a8deb